Assets/Scripts/ProceduralGeneration.cs

```diff
@@ -0,0 +1,192 @@
+﻿using System.Collections;
+using UnityEngine;
+using System.Collections.Generic;
+using Quaternion = UnityEngine.Quaternion;
+using Random = UnityEngine.Random;
+using Vector3 = UnityEngine.Vector3;
+
+
+public class ProceduralGeneration : MonoBehaviour
+{
+    // Unit cubes should evenly subdivide chunk, and chunks should evenly subdivide level.
+    public Vector3Int levelDimensions;
+    public int chunkSize;
+    public int unitCubeSize;
+    public float asteroidDensity = 0.5f;
+    public Vector2 asteroidSizeRange;
+    public float viewDistance = 500f;
+    
+    // public GameObject blackHolePrefab;
+    // public GameObject exitPortalPrefab;
+    public GameObject asteroidPrefab;
+
+    private GameObject asteroidsCollection;
+    
+    private Vector3Int numChunksInLevel;
+    private Vector3Int numUnitCubesInChunk;
+
+    private CullingGroup cullGroup;
+    private BoundingSphere[] bounds;
+    private Dictionary<int, GameObject> entityLookup;
+    private int EntityCount => entityLookup.Count;
+
+    private Dictionary<int, Transform> entityTransforms;
+    
+    private Dictionary<int, int> entityIdToIndex;
+    private Dictionary<int, int> entityIndexToId;
+    
+    public static bool FinishedGenerating = false;
+    private int maxEntities;
+
+    private void Awake()
+    {
+        asteroidsCollection = GameObject.FindGameObjectWithTag("Asteroids");
+
+        numChunksInLevel = levelDimensions / chunkSize;
+        numUnitCubesInChunk = Vector3Int.one * (chunkSize / unitCubeSize);
+
+        // var asteroidsPerChunk = 1; TODO: Make this more extendable
+        maxEntities = numChunksInLevel.x * numUnitCubesInChunk.x * numChunksInLevel.y * numUnitCubesInChunk.y * numChunksInLevel.z * numUnitCubesInChunk.z  * 1; // 1 asteroid per unit cube right now
+        Debug.Log($"Max Entities: {maxEntities}");
+        // Culling Setup
+        var mainCam = Camera.main;
+        cullGroup = new CullingGroup {targetCamera = mainCam};
+        cullGroup.SetDistanceReferencePoint(mainCam.transform);
+        cullGroup.SetBoundingDistances(new[] {viewDistance, float.PositiveInfinity});
+        bounds = new BoundingSphere[maxEntities];
+        cullGroup.SetBoundingSpheres(bounds);
+        cullGroup.SetBoundingSphereCount(0);
+        cullGroup.onStateChanged += CullingStateChanged;
+        entityLookup = new Dictionary<int, GameObject>();
+        entityTransforms = new Dictionary<int, Transform>();
+        entityIdToIndex = new Dictionary<int, int>();
+        entityIndexToId = new Dictionary<int, int>();
+
+        StartCoroutine(nameof(GenerateAsteroidField));
+        StartCoroutine(nameof(UpdateCulledObjectBounds));
+    }
+
+    public void AddEntity(GameObject obj)
```

Comment:
Can we make this private until it needs to be public?

Reply:
Will do